src/tribler/gui/network/request_manager.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import annotations
+
+import json
+import logging
+from time import time
+from typing import Dict, TYPE_CHECKING
+from urllib.parse import quote_plus
+
+from PyQt5.QtCore import QBuffer, QIODevice, QUrl
+from PyQt5.QtNetwork import QNetworkAccessManager, QNetworkRequest
+
+from tribler.core.utilities.limited_ordered_dict import LimitedOrderedDict
+from tribler.gui.defs import BUTTON_TYPE_NORMAL, DEFAULT_API_HOST, DEFAULT_API_PORT, DEFAULT_API_PROTOCOL
+from tribler.gui.dialogs.confirmationdialog import ConfirmationDialog
+from tribler.gui.network.request.shutdown_request import ShutdownRequest
+from tribler.gui.utilities import connect
+
+if TYPE_CHECKING:
+    from tribler.gui.network.request.request import Request
+
+
+class RequestManager(QNetworkAccessManager):
+    """
+    This class is responsible for all the requests made to the Tribler REST API.
+    All requests are asynchronous so the caller object should keep track of response (QNetworkReply) object. A finished
+    pyqt signal is fired when the response data is ready.
+    """
+
+    window = None
+
+    def __init__(self, limit: int = 50, timeout_interval: int = 15):
+        QNetworkAccessManager.__init__(self)
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+        self.active_requests = {}
+        self.performed_requests: Dict[Request, int] = LimitedOrderedDict(limit=200)
+
+        self.protocol = DEFAULT_API_PROTOCOL
+        self.host = DEFAULT_API_HOST
+        self.port = DEFAULT_API_PORT
+        self.key = b""
+        self.limit = limit
+        self.timeout_interval = timeout_interval
+
+    def add(self, request: Request):
+        if len(self.active_requests) > self.limit:
+            self._drop_timed_out_requests()
+
+        self.active_requests[request] = request
+        self.performed_requests[request] = 0
+        request.manager = self
+        request.url = self._get_base_url() + request.endpoint
+        request.url += f'?{self._urlencode(request.url_params)}' if request.url_params else ''
+        self.logger.info(f'Request: {request}')
```

Comment:
It looks a bit nicer if the manager is assigned by calling a new `request.set_manager(...)` method. In this case, it becomes possible to move the URL-handling logic to the Request class. It looks clearer if the `Request` class itself handles its `url` attribute.